pkg/storage/snapshot/source.go

```diff
@@ -49,65 +54,129 @@ const (
 	sourceFinalizer = "snapshot.kubevirt.io/snapshot-source-protection"
 )
 
+var (
+	ErrVolumeDoesntExist  = errors.New("volume doesnt exist")
+	ErrVolumeNotBound     = errors.New("volume not bound")
+	ErrVolumeNotPopulated = errors.New("volume not populated")
+)
+
 type snapshotSource interface {
+	UpdateSourceState() error
 	UID() types.UID
 	Locked() bool
+	LockMsg() string
 	Lock() (bool, error)
 	Unlock() (bool, error)
-	Online() (bool, error)
-	GuestAgent() (bool, error)
-	Frozen() (bool, error)
+	Online() bool
+	GuestAgent() bool
+	Frozen() bool
 	Freeze() error
 	Unfreeze() error
 	Spec() (snapshotv1.SourceSpec, error)
 	PersistentVolumeClaims() (map[string]string, error)
 }
 
+type sourceState struct {
+	online     bool
+	guestAgent bool
+	frozen     bool
+	locked     bool
+	lockMsg    string
+}
+
 type vmSnapshotSource struct {
 	vm         *kubevirtv1.VirtualMachine
 	snapshot   *snapshotv1.VirtualMachineSnapshot
 	controller *VMSnapshotController
+	state      *sourceState
+}
+
+func (s *vmSnapshotSource) UpdateSourceState() error {
+	vmi, exists, err := s.controller.getVMI(s.vm)
+	if err != nil {
+		return err
+	}
+
+	online := exists
+
```

Comment:
What iv vmi is nil?

Reply:
not sure what is the question here?
The code to get online was:
```
func (ctrl *VMSnapshotController) checkVMIRunning(vm *kubevirtv1.VirtualMachine) (bool, error) {
	_, exists, err := ctrl.getVMI(vm)
	return exists, err
}
func (s *vmSnapshotSource) Online() (bool, error) {
	exists, err := s.controller.checkVMIRunning(s.vm)
	if err != nil {
		return false, err
	}

	return exists, nil
}
```
which is exactly what I did here, and anywhere that uses the vmi checks first `exists` which if exists is true then vmi is not nil.. if exists is false then online is false as expected.. so not sure whats wrong..

Reply:
You're accessing the `vmi` fields regardless of the `online/exists` in this func

Reply:
my bad it's all guarded with `exists &&`

Reply:
I always check `exists &&` before any use of the vmi